src/Microsoft.Diagnostics.Monitoring.EventPipe/Counters/TraceEventExtensions.cs

```diff
@@ -29,8 +29,21 @@ public CounterConfiguration(CounterFilter filter)
 
     internal static class TraceEventExtensions
     {
+        private static Dictionary<string, Provider> providersByName = new();
         private static HashSet<string> inactiveSharedSessions = new(StringComparer.OrdinalIgnoreCase);
 
+        // This assumes uniqueness of provider names - this is currently a limitation (see https://github.com/dotnet/runtime/issues/93097)
+        public static Provider GetOrCreateProvider(string providerName, string meterTags = null, string instrumentTags = null, string scopeHash = null)
+        {
+            if (providersByName.TryGetValue(providerName, out Provider provider))
```

Comment:
InstrumentTags vary per-instrument rather than per-provider. Currently the code appears to cache whatever instrument tags occur in the first instrument created for a Meter and then those tags are used incorrectly for all other instruments on the same Meter. I think it would be more accurate rename Provider to 'CachedCounterInfo' and the lookup key needs to be a tuple of provider name + counter name.

Reply:
Thanks for pointing this out, that was an oversight on my part in the initial version. The latest version should handle this correctly (assuming I understood what you're describing), and I added more tests to validate that scenario.